File: internal/azure/cert.go
package azure

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
	"time"
)

// CertificateManager handles Azure AD SSH certificate lifecycle and renewal
type CertificateManager struct {
	// Map of VM name to certificate info
	certs map[string]*CertInfo
	mu    sync.RWMutex

	// Status channel for certificate events
	statusCh chan CertStatus

	// Context for cancellation
	ctx    context.Context
	cancel context.CancelFunc
}

// CertInfo stores information about a certificate
type CertInfo struct {
	VMName         string
	SSHConfigPath  string // Path to SSH config directory (e.g., ~/.ssh/az_ssh_config/vm-name)
	PublicKeyPath  string // Path to public key
	CertPath       string // Path to certificate file
	CreatedAt      time.Time
	ExpiresAt      time.Time
	LastRenewalTry time.Time
	RenewalStatus  string // "valid", "expiring_soon", "expired", "renewing", "renewal_failed"
	mu             sync.RWMutex
}

// CertStatus represents a certificate status update
type CertStatus struct {
	VMName    string
	Status    string // "valid", "renewed", "expiring_soon", "expired", "renewal_failed"
	Message   string
	ExpiresIn time.Duration
}

const (
	CertLifetime      = 1 * time.Hour    // Azure AD SSH certs expire after 1 hour
	RenewalWindow     = 5 * time.Minute  // Start trying to renew 5 minutes before expiry
	RenewalRetryDelay = 30 * time.Second // Retry every 30 seconds if renewal fails
	CheckInterval     = 1 * time.Minute  // Check certificate status every minute
)

// NewCertificateManager creates a new certificate manager
func NewCertificateManager() *CertificateManager {
	ctx, cancel := context.WithCancel(context.Background())
	return &CertificateManager{
		certs:    make(map[string]*CertInfo),
		statusCh: make(chan CertStatus, 10),
		ctx:      ctx,
		cancel:   cancel,
	}
}

// RegisterCertificate registers a certificate for monitoring and auto-renewal
// sshConfigPath should be the base path for SSH config (e.g., ~/.ssh/az_ssh_config/vm-name)
func (cm *CertificateManager) RegisterCertificate(vmName, sshConfigPath string) error {
	cm.mu.Lock()
	defer cm.mu.Unlock()

	// Expand home directory if needed
	if strings.HasPrefix(sshConfigPath, "~") {
		homeDir, err := os.UserHomeDir()
		if err != nil {
			return fmt.Errorf("failed to get home directory: %w", err)
		}
		sshConfigPath = filepath.Join(homeDir, sshConfigPath[2:])
	}

	// Construct paths
	publicKeyPath := filepath.Join(sshConfigPath, "id_rsa.pub")
	certPath := publicKeyPath + "-aadcert.pub"

	// Check if certificate exists
	_, err := os.Stat(certPath)
	if err != nil {
		if os.IsNotExist(err) {
			// Certificate doesn't exist yet - mark as expired so it will be generated
			cm.certs[vmName] = &CertInfo{
				VMName:        vmName,
				SSHConfigPath: sshConfigPath,
				PublicKeyPath: publicKeyPath,
				CertPath:      certPath,
				ExpiresAt:     time.Now(), // Already expired
				RenewalStatus: "expired",
			}
			return nil
		}
		return fmt.Errorf("failed to check certificate: %w", err)
	}

	// Try to read the certificate expiry time using ssh-keygen
	var expiresAt time.Time
	cmd := exec.Command("ssh-keygen", "-L", "-f", certPath)
	output, err := cmd.CombinedOutput()
	if err == nil {
		// Parse the certificate details
		expiresAt, err = parseCertificateExpiry(string(output))
		if err != nil {
			// Fallback to file modification time + 1 hour
			certInfo, _ := os.Stat(certPath)
			expiresAt = certInfo.ModTime().Add(CertLifetime)
		}
	} else {
		// Fallback to file modification time + 1 hour
		certInfo, _ := os.Stat(certPath)
		expiresAt = certInfo.ModTime().Add(CertLifetime)
	}

	ci := &CertInfo{
		VMName:        vmName,
		SSHConfigPath: sshConfigPath,
		PublicKeyPath: publicKeyPath,
		CertPath:      certPath,
		CreatedAt:     time.Now(),
		ExpiresAt:     expiresAt,
		RenewalStatus: getRenewalStatus(expiresAt),
	}

	cm.certs[vmName] = ci

	// Send initial status
	select {
	case cm.statusCh <- CertStatus{
		VMName:    vmName,
		Status:    ci.RenewalStatus,
		Message:   fmt.Sprintf("Certificate tracked: expires at %s", expiresAt.Format("15:04:05")),
		ExpiresIn: time.Until(expiresAt),
	}:
	default:
	}

	return nil
}

// StartMonitoring starts the certificate monitoring and auto-renewal goroutine
func (cm *CertificateManager) StartMonitoring() {
	go cm.monitorLoop()
}

// monitorLoop continuously monitors certificates and handles renewal
func (cm *CertificateManager) monitorLoop() {
	ticker := time.NewTicker(CheckInterval)
	defer ticker.Stop()

	for {
		select {
		case <-cm.ctx.Done():
			return
		case <-ticker.C:
			cm.checkAndRenewCertificates()
		}
	}
}

// checkAndRenewCertificates checks all registered certificates and renews if needed
func (cm *CertificateManager) checkAndRenewCertificates() {
	cm.mu.RLock()
	certs := make([]*CertInfo, 0, len(cm.certs))
	for _, cert := range cm.certs {
		certs = append(certs, cert)
	}
	cm.mu.RUnlock()

	now := time.Now()

	for _, cert := range certs {
		cert.mu.Lock()
		expiresAt := cert.ExpiresAt
		lastTry := cert.LastRenewalTry
		status := cert.RenewalStatus
		cert.mu.Unlock()

		timeUntilExpiry := time.Until(expiresAt)

		// Update status based on time until expiry
		newStatus := getRenewalStatus(expiresAt)
		if newStatus != status {
			cert.mu.Lock()
			cert.RenewalStatus = newStatus
			cert.mu.Unlock()

			select {
			case cm.statusCh <- CertStatus{
				VMName:    cert.VMName,
				Status:    newStatus,
				Message:   fmt.Sprintf("Certificate status changed to: %s", newStatus),
				ExpiresIn: timeUntilExpiry,
			}:
			default:
			}
		}

		// Check if we should attempt renewal
		// Renewal window: last 5 minutes before expiry
		// But also try immediately if expired or in renewal_failed state
		shouldRenew := false

		if timeUntilExpiry <= 0 {
			// Certificate expired - try to renew
			shouldRenew = true
		} else if timeUntilExpiry <= RenewalWindow {
			// Within renewal window - try to renew
			// Only retry if we haven't tried in the last 30 seconds
			if now.Sub(lastTry) >= RenewalRetryDelay {
				shouldRenew = true
			}
		}

		if shouldRenew {
			go cm.renewCertificate(cert)
		}
	}
}

// renewCertificate attempts to renew a certificate
func (cm *CertificateManager) renewCertificate(cert *CertInfo) {
	cert.mu.Lock()
	cert.LastRenewalTry = time.Now()
	cert.RenewalStatus = "renewing"
	vmName := cert.VMName
	publicKeyPath := cert.PublicKeyPath
	certPath := cert.CertPath
	cert.mu.Unlock()

	// Notify that renewal is starting
	select {
	case cm.statusCh <- CertStatus{
		VMName:  vmName,
		Status:  "renewing",
		Message: "Attempting to renew certificate...",
	}:
	default:
	}

	// Run: az ssh cert --file <cert-path> --public-key-file <public-key-path>
	cmd := exec.CommandContext(cm.ctx, "az", "ssh", "cert",
		"--file", certPath,
		"--public-key-file", publicKeyPath,
	)

	output, err := cmd.CombinedOutput()
	if err != nil {
		cert.mu.Lock()
		cert.RenewalStatus = "renewal_failed"
		cert.mu.Unlock()

		errMsg := string(output)
		if errMsg == "" {
			errMsg = err.Error()
		}

		select {
		case cm.statusCh <- CertStatus{
			VMName:  vmName,
			Status:  "renewal_failed",
			Message: fmt.Sprintf("Certificate renewal failed: %s", errMsg),
		}:
		default:
		}
		return
	}

	// Parse the actual expiry time from the output
	expiresAt, err := parseExpiryFromOutput(string(output))
	if err != nil {
		// Fallback to estimated time if parsing fails
		expiresAt = time.Now().Add(CertLifetime)
	}

	cert.mu.Lock()
	cert.CreatedAt = time.Now()
	cert.ExpiresAt = expiresAt
	cert.RenewalStatus = "valid"
	cert.mu.Unlock()

	select {
	case cm.statusCh <- CertStatus{
		VMName:    vmName,
		Status:    "renewed",
		Message:   fmt.Sprintf("Certificate renewed successfully! Expires at: %s", expiresAt.Format("15:04:05")),
		ExpiresIn: time.Until(expiresAt),
	}:
	default:
	}
}

// GenerateCertificate generates a new certificate for the first time
// This should be called before the first connection
func (cm *CertificateManager) GenerateCertificate(vmName, sshConfigPath, publicKeyPath string) error {
	// Expand home directory if needed
	if strings.HasPrefix(sshConfigPath, "~") {
		homeDir, err := os.UserHomeDir()
		if err != nil {
			return fmt.Errorf("failed to get home directory: %w", err)
		}
		sshConfigPath = filepath.Join(homeDir, sshConfigPath[2:])
	}

	if strings.HasPrefix(publicKeyPath, "~") {
		homeDir, err := os.UserHomeDir()
		if err != nil {
			return fmt.Errorf("failed to get home directory: %w", err)
		}
		publicKeyPath = filepath.Join(homeDir, publicKeyPath[2:])
	}

	certPath := publicKeyPath + "-aadcert.pub"

	// Ensure directory exists
	if err := os.MkdirAll(filepath.Dir(publicKeyPath), 0o700); err != nil {
		return fmt.Errorf("failed to create SSH config directory: %w", err)
	}

	// Generate SSH key pair if it doesn't exist
	if _, err := os.Stat(publicKeyPath); os.IsNotExist(err) {
		privateKeyPath := strings.TrimSuffix(publicKeyPath, ".pub")
		cmd := exec.Command("ssh-keygen", "-t", "rsa", "-b", "4096",
			"-f", privateKeyPath, "-N", "")
		if output, err := cmd.CombinedOutput(); err != nil {
			return fmt.Errorf("failed to generate SSH key: %s: %w", string(output), err)
		}
	}

	// Generate the Azure AD certificate
	cmd := exec.CommandContext(cm.ctx, "az", "ssh", "cert",
		"--file", certPath,
		"--public-key-file", publicKeyPath,
	)

	output, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("failed to generate certificate: %s: %w", string(output), err)
	}

	// Parse the actual expiry time from the output
	expiresAt, err := parseExpiryFromOutput(string(output))
	if err != nil {
		// Fallback to estimated time if parsing fails
		expiresAt = time.Now().Add(CertLifetime)
	}

	// Register the certificate for monitoring with the actual expiry time
	cm.mu.Lock()
	cm.certs[vmName] = &CertInfo{
		VMName:        vmName,
		SSHConfigPath: sshConfigPath,
		PublicKeyPath: publicKeyPath,
		CertPath:      certPath,
		CreatedAt:     time.Now(),
		ExpiresAt:     expiresAt,
		RenewalStatus: "valid",
	}
	cm.mu.Unlock()

	// Send status update
	select {
	case cm.statusCh <- CertStatus{
		VMName:    vmName,
		Status:    "valid",
		Message:   fmt.Sprintf("Certificate generated! Expires at: %s", expiresAt.Format("15:04:05")),
		ExpiresIn: time.Until(expiresAt),
	}:
	default:
	}

	return nil
}

// GetStatus returns the current status of a certificate
func (cm *CertificateManager) GetStatus(vmName string) (string, time.Duration, error) {
	cm.mu.RLock()
	cert, exists := cm.certs[vmName]
	cm.mu.RUnlock()

	if !exists {
		return "", 0, fmt.Errorf("certificate not registered")
	}

	cert.mu.RLock()
	defer cert.mu.RUnlock()

	return cert.RenewalStatus, time.Until(cert.ExpiresAt), nil
}

// StatusChannel returns the channel for certificate status updates
func (cm *CertificateManager) StatusChannel() <-chan CertStatus {
	return cm.statusCh
}

// Stop stops the certificate manager
func (cm *CertificateManager) Stop() {
	cm.cancel()
}

// getRenewalStatus determines the renewal status based on time until expiry
func getRenewalStatus(expiresAt time.Time) string {
	timeUntil := time.Until(expiresAt)

	if timeUntil <= 0 {
		return "expired"
	} else if timeUntil <= RenewalWindow {
		return "expiring_soon"
	}
	return "valid"
}

// parseExpiryFromOutput parses the expiry time from Azure CLI output
// Example output: "Generated SSH certificate ... is valid until 2025-10-15 18:06:23 in local time."
func parseExpiryFromOutput(output string) (time.Time, error) {
	// Match pattern: "is valid until YYYY-MM-DD HH:MM:SS in local time"
	re := regexp.MustCompile(`is valid until (\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}) in local time`)
	matches := re.FindStringSubmatch(output)

	if len(matches) < 2 {
		return time.Time{}, fmt.Errorf("could not parse expiry time from output")
	}

	// Parse the timestamp
	expiryStr := matches[1]
	expiresAt, err := time.ParseInLocation("2006-01-02 15:04:05", expiryStr, time.Local)
	if err != nil {
		return time.Time{}, fmt.Errorf("failed to parse expiry time: %w", err)
	}

	return expiresAt, nil
}

// parseCertificateExpiry parses expiry time from ssh-keygen -L output
// Example output contains: "Valid: from 2025-10-15T17:31:23 to 2025-10-15T18:31:23"
func parseCertificateExpiry(output string) (time.Time, error) {
	// Match pattern: "Valid: from ... to YYYY-MM-DDTHH:MM:SS"
	re := regexp.MustCompile(`Valid: from .+ to (\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})`)
	matches := re.FindStringSubmatch(output)

	if len(matches) < 2 {
		return time.Time{}, fmt.Errorf("could not parse certificate expiry from ssh-keygen output")
	}

	// Parse the timestamp (it's in local time)
	expiryStr := matches[1]
	expiresAt, err := time.ParseInLocation("2006-01-02T15:04:05", expiryStr, time.Local)
	if err != nil {
		return time.Time{}, fmt.Errorf("failed to parse certificate expiry: %w", err)
	}

	return expiresAt, nil
}

File: src/azure/cert.rs
use crate::azure::parse::{parse_certificate_expiry, parse_expiry_from_output};
use crate::config::expand_tilde;
use crate::model::CertStatus;
use crate::tui::action::BgEvent;
use chrono::{DateTime, Duration as ChronoDuration, Local};
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::{Arc, Mutex};
use std::time::Duration;
use tokio::process::Command;
use tokio::sync::mpsc::UnboundedSender;

const CERT_LIFETIME: ChronoDuration = ChronoDuration::hours(1);
const RENEWAL_WINDOW_MINS: i64 = 5;
const RENEWAL_RETRY: ChronoDuration = ChronoDuration::seconds(30);
const CHECK_INTERVAL: Duration = Duration::from_secs(60);

#[derive(Debug, Clone)]
struct CertInfo {
    vm_name: String,
    public_key_path: PathBuf,
    cert_path: PathBuf,
    expires_at: DateTime<Local>,
    last_renewal_try: Option<DateTime<Local>>,
    status: CertStatus,
}

/// Determine status from expiry, matching Go getRenewalStatus.
fn renewal_status(expires_at: DateTime<Local>) -> CertStatus {
    let remaining = expires_at - Local::now();
    if remaining <= ChronoDuration::zero() {
        CertStatus::Expired
    } else if remaining <= ChronoDuration::minutes(RENEWAL_WINDOW_MINS) {
        CertStatus::ExpiringSoon
    } else {
        CertStatus::Valid
    }
}

#[derive(Clone)]
pub struct CertManager {
    tx: UnboundedSender<BgEvent>,
    certs: Arc<Mutex<HashMap<String, CertInfo>>>,
}

impl CertManager {
    pub fn new(tx: UnboundedSender<BgEvent>) -> Self {
        Self { tx, certs: Arc::new(Mutex::new(HashMap::new())) }
    }

    /// Register a cert for monitoring (cert may not exist yet -> marked expired).
    pub fn register(&self, vm_name: &str, ssh_config_path: &str) {
        let dir = PathBuf::from(expand_tilde(ssh_config_path));
        let public_key_path = dir.join("id_rsa.pub");
        let cert_path = dir.join("id_rsa.pub-aadcert.pub");

        let (expires_at, status) = if cert_path.exists() {
            let exp = read_cert_expiry(&cert_path).unwrap_or_else(|| Local::now() + CERT_LIFETIME);
            (exp, renewal_status(exp))
        } else {
            (Local::now(), CertStatus::Expired)
        };

        let info = CertInfo {
            vm_name: vm_name.to_string(),
            public_key_path,
            cert_path,
            expires_at,
            last_renewal_try: None,
            status,
        };
        let expires_in = (info.expires_at - Local::now()).to_std().ok();
        self.certs.lock().unwrap().insert(vm_name.to_string(), info);
        let _ = self.tx.send(BgEvent::Cert { vm_name: vm_name.to_string(), status, expires_in });
    }

    /// Spawn the periodic check-and-renew loop.
    pub fn start_monitoring(&self) {
        let me = self.clone();
        tokio::spawn(async move {
            let mut ticker = tokio::time::interval(CHECK_INTERVAL);
            loop {
                ticker.tick().await;
                me.check_and_renew().await;
            }
        });
    }

    async fn check_and_renew(&self) {
        let snapshot: Vec<CertInfo> = self.certs.lock().unwrap().values().cloned().collect();
        let now = Local::now();
        for cert in snapshot {
            let new_status = renewal_status(cert.expires_at);
            if new_status != cert.status {
                if let Some(c) = self.certs.lock().unwrap().get_mut(&cert.vm_name) {
                    c.status = new_status;
                }
                let expires_in = (cert.expires_at - now).to_std().ok();
                let _ = self.tx.send(BgEvent::Cert { vm_name: cert.vm_name.clone(), status: new_status, expires_in });
            }

            let remaining = cert.expires_at - now;
            let should_renew = remaining <= ChronoDuration::zero()
                || (remaining <= ChronoDuration::minutes(RENEWAL_WINDOW_MINS)
                    && cert.last_renewal_try.is_none_or(|t| now - t >= RENEWAL_RETRY));
            if should_renew {
                self.renew(cert.vm_name.clone()).await;
            }
        }
    }

    async fn renew(&self, vm_name: String) {
        let (public_key_path, cert_path) = {
            let mut guard = self.certs.lock().unwrap();
            let Some(c) = guard.get_mut(&vm_name) else { return };
            c.last_renewal_try = Some(Local::now());
            c.status = CertStatus::Renewing;
            (c.public_key_path.clone(), c.cert_path.clone())
        };
        let _ = self.tx.send(BgEvent::Cert { vm_name: vm_name.clone(), status: CertStatus::Renewing, expires_in: None });

        let output = Command::new("az")
            .arg("ssh").arg("cert")
            .arg("--file").arg(&cert_path)
            .arg("--public-key-file").arg(&public_key_path)
            .output().await;

        match output {
            Ok(out) if out.status.success() => {
                let text = String::from_utf8_lossy(&out.stdout);
                let expires_at = parse_expiry_from_output(&text).unwrap_or_else(|_| Local::now() + CERT_LIFETIME);
                if let Some(c) = self.certs.lock().unwrap().get_mut(&vm_name) {
                    c.expires_at = expires_at;
                    c.status = CertStatus::Valid;
                }
                let expires_in = (expires_at - Local::now()).to_std().ok();
                let _ = self.tx.send(BgEvent::Cert { vm_name, status: CertStatus::Renewed, expires_in });
            }
            _ => {
                // Renewal failed (az error or non-zero exit). We surface this only as
                // the RenewalFailed status, matching the Go TUI, which likewise does not
                // display the underlying error message. A diagnostic log file is Phase 2.
                if let Some(c) = self.certs.lock().unwrap().get_mut(&vm_name) {
                    c.status = CertStatus::RenewalFailed;
                }
                let _ = self.tx.send(BgEvent::Cert { vm_name, status: CertStatus::RenewalFailed, expires_in: None });
            }
        }
    }

    /// Manual (re)generation triggered by `r`. Runs ssh-keygen if no key, then az ssh cert.
    pub async fn generate(&self, vm_name: String, ssh_config_path: String) {
        let dir = PathBuf::from(expand_tilde(&ssh_config_path));
        let public_key_path = dir.join("id_rsa.pub");
        let private_key_path = dir.join("id_rsa");
        let cert_path = dir.join("id_rsa.pub-aadcert.pub");

        if let Err(e) = std::fs::create_dir_all(&dir) {
            let _ = self.tx.send(BgEvent::CertRegenResult { vm_name, ok: false, message: format!("mkdir failed: {e}") });
            return;
        }

        if !public_key_path.exists() {
            let kg = Command::new("ssh-keygen")
                .arg("-t").arg("rsa").arg("-b").arg("4096")
                .arg("-f").arg(&private_key_path).arg("-N").arg("")
                .output().await;
            if let Ok(out) = &kg {
                if !out.status.success() {
                    let _ = self.tx.send(BgEvent::CertRegenResult { vm_name, ok: false, message: String::from_utf8_lossy(&out.stderr).to_string() });
                    return;
                }
            } else if let Err(e) = kg {
                let _ = self.tx.send(BgEvent::CertRegenResult { vm_name, ok: false, message: e.to_string() });
                return;
            }
        }

        let out = Command::new("az")
            .arg("ssh").arg("cert")
            .arg("--file").arg(&cert_path)
            .arg("--public-key-file").arg(&public_key_path)
            .output().await;

        match out {
            Ok(o) if o.status.success() => {
                let text = String::from_utf8_lossy(&o.stdout);
                let expires_at = parse_expiry_from_output(&text).unwrap_or_else(|_| Local::now() + CERT_LIFETIME);
                self.certs.lock().unwrap().insert(vm_name.clone(), CertInfo {
                    vm_name: vm_name.clone(),
                    public_key_path,
                    cert_path,
                    expires_at,
                    last_renewal_try: None,
                    status: CertStatus::Valid,
                });
                let expires_in = (expires_at - Local::now()).to_std().ok();
                let _ = self.tx.send(BgEvent::Cert { vm_name: vm_name.clone(), status: CertStatus::Valid, expires_in });
                let _ = self.tx.send(BgEvent::CertRegenResult { vm_name, ok: true, message: "Certificate regenerated".into() });
            }
            other => {
                let msg = match other {
                    Ok(o) => String::from_utf8_lossy(&o.stderr).to_string(),
                    Err(e) => e.to_string(),
                };
                let _ = self.tx.send(BgEvent::CertRegenResult { vm_name, ok: false, message: msg });
            }
        }
    }
}

/// Read cert expiry via `ssh-keygen -L -f <cert>`, falling back to file mtime + 1h.
fn read_cert_expiry(cert_path: &std::path::Path) -> Option<DateTime<Local>> {
    let out = std::process::Command::new("ssh-keygen").arg("-L").arg("-f").arg(cert_path).output().ok()?;
    let text = String::from_utf8_lossy(&out.stdout);
    if let Ok(exp) = parse_certificate_expiry(&text) {
        return Some(exp);
    }
    let meta = std::fs::metadata(cert_path).ok()?;
    let modified: DateTime<Local> = meta.modified().ok()?.into();
    Some(modified + CERT_LIFETIME)
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Duration as ChronoDuration;

    #[test]
    fn status_expired_when_past() {
        let exp = chrono::Local::now() - ChronoDuration::minutes(1);
        assert_eq!(renewal_status(exp), crate::model::CertStatus::Expired);
    }

    #[test]
    fn status_expiring_within_window() {
        let exp = chrono::Local::now() + ChronoDuration::minutes(3);
        assert_eq!(renewal_status(exp), crate::model::CertStatus::ExpiringSoon);
    }

    #[test]
    fn status_valid_when_far() {
        let exp = chrono::Local::now() + ChronoDuration::minutes(50);
        assert_eq!(renewal_status(exp), crate::model::CertStatus::Valid);
    }
}
